### PR TITLE
JNI bindings for distinct_hash_join

### PR DESCRIPTION
**Pull request by jlowe**
## Description
Adds Java bindings to the distinct hash join functionality added in #14990.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Review by jbrennan333 — APPROVED** [review]
This looks good (other than minor checkstyle issues).

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 1]
```diff
@@ -3150,6 +3153,28 @@ public GatherMap[] innerJoinGatherMaps(Table rightKeys, boolean compareNullsEqua
...
 
+  /**
+   * Computes the gather maps that can be used to manifest the result of an inner equi-join between
+   * two tables where the right table is guaranteed to not contain any duplicated join keys. It is
+   * assumed this table instance holds the key columns from the left table, and the table argument
+   * represents the key columns from the right table. Two {@link GatherMap} instances will be
+   * returned that can be used to gather the left and right tables, respectively, to produce the
+   * result of the inner join.
+   * It is the responsibility of the caller to close the resulting gather map instances.
+   * @param rightKeys join key columns from the right table
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:3164]:
```suggestion
   * result of the inner join.
   *
   * It is the responsibility of the caller to close the resulting gather map instances.
   *
   * @param rightKeys join key columns from the right table
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 2]
```diff
@@ -3150,6 +3153,28 @@ public GatherMap[] innerJoinGatherMaps(Table rightKeys, boolean compareNullsEqua
...
+   * @param compareNullsEqual true if null key values should match otherwise false
+   * @return left and right table gather maps
+   */
+  public GatherMap[] innerDistinctJoinGatherMaps(Table rightKeys, boolean compareNullsEqual) {
+    if (getNumberOfColumns() != rightKeys.getNumberOfColumns()) {
+      throw new IllegalArgumentException("column count mismatch, this: " + getNumberOfColumns() +
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:3170]:
```suggestion
      throw new IllegalArgumentException("Column count mismatch, this: " + getNumberOfColumns() +
```

File: java/src/main/native/src/TableJni.cpp [hunk 3]
```diff
@@ -2550,6 +2550,30 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_innerJoinGatherMaps(
...
+      env, j_left_keys, j_right_keys, compare_nulls_equal,
+      [](cudf::table_view const &left, cudf::table_view const &right, cudf::null_equality nulleq) {
+        auto has_nulls = cudf::has_nested_nulls(left) || cudf::has_nested_nulls(right) ?
+                             cudf::nullable_join::YES :
+                             cudf::nullable_join::NO;
+        std::pair<std::unique_ptr<rmm::device_uvector<cudf::size_type>>,
+                  std::unique_ptr<rmm::device_uvector<cudf::size_type>>>
+            maps;
+        if (cudf::detail::has_nested_columns(right)) {
+          cudf::distinct_hash_join<cudf::has_nested::YES> hash(right, left, has_nulls, nulleq);
```
Comment (ttnghia) [inline java/src/main/native/src/TableJni.cpp:2565]: Why do we call `distinct_hash_join<cudf::has_nested::YES>` without considering the left table?

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 4]
```diff
@@ -2085,6 +2084,116 @@ void testInnerJoinGatherMapsNulls() {
...
+  private void checkInnerDistinctJoin(Table leftKeys, Table rightKeys, Table expected,
+                                      boolean compareNullsEqual) {
+    GatherMap[] maps = leftKeys.innerDistinctJoinGatherMaps(rightKeys, compareNullsEqual);
+    try {
+      verifyJoinGatherMaps(maps, expected);
+    } finally {
+      for (GatherMap map : maps) {
+        map.close();
+      }
+    }
```
Comment (ttnghia) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:2096]: Can we implement `GatherMap` as auto-closable?

Comment (jlowe) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:2096]: GatherMap is auto-closeable.  This is an array of GatherMaps.

Comment (ttnghia) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:2096]: Hah, array of auto-closeable can't be auto-closeable. That sucks.

**Review by ttnghia — APPROVED** [review]

**Comment by raydouglass** [comment]
Admin merging while Python 3.11 is sorted out.

**Pull request merged**